src/databricks/sql/utils.py

```diff
@@ -542,20 +542,41 @@ def infer_types(params: list[DbSqlParameter]):
         datetime.date: DbSqlType.DATE,
         bool: DbSqlType.BOOLEAN,
         Decimal: DbSqlType.DECIMAL,
+        type(None): DbSqlType.VOID,
     }
-    new_params = copy.deepcopy(params)
-    for param in new_params:
-        if not param.type:
-            if type(param.value) in type_lookup_table:
-                param.type = type_lookup_table[type(param.value)]
-            else:
-                raise ValueError("Parameter type cannot be inferred")
-
-        if param.type == DbSqlType.DECIMAL:
+
+    new_params = []
+
+    # cycle through each parameter we've been passed
+    for param in params:
+        _name: str = param.name
```

Comment:
Strive for each unit of code having a single reason to change (Single Responsibility Principle).  E.g., here I would, at a minimum, suggest that the logic inside the loop be a separate function, but might go so far as to have the special logic for each type in their own methods, so that you end up with

def infer_types(params: list[DbSqlParameter]) -> list[DbSqlParameter]
def infer_type(param: DbSqlParameter) -> DbSqlParameter
def _convert_to_decimal_type(decimal_parameter: DbSqlParameter) -> DbSqlDynamicDecimalType
def _conver_to_void_type(decimal_parameter: DbSqlParameter) -> DbSqlVoidType

where infer_type has a switch statement (or if/elif I guess, stupid old pythons).

Reply:
Ideally, most functions in a dynamic language like python should be 10 lines or less, to improve comprehension and make testing easier (less likely to have to test combinations of logical branches together).